TombEngine/Specific/Video/Video.cpp

```diff
@@ -0,0 +1,377 @@
+	{
+		if (_videoTexture || _textureView)
+		{
+			TENLog("Video texture already exists", LogLevel::Error);
+			return false;
+		}
+
+		auto resolution = g_Renderer.GetScreenResolution();
+		
+		// Limit maximum resolution, because some GPUs may not handle textures larger than 2048x2048.
+		_textureSize.x = std::clamp(resolution.x, MIN_VIDEO_WIDTH, MAX_VIDEO_WIDTH);
+		_textureSize.y = std::clamp(resolution.y, MIN_VIDEO_HEIGHT, MAX_VIDEO_HEIGHT);
+
+		libvlc_video_set_format(_player, "BGRA", _textureSize.x, _textureSize.y, _textureSize.x * 4);
+
+		_frameBuffer.resize(_textureSize.x * _textureSize.y * 4);
+
+		D3D11_TEXTURE2D_DESC texDesc = {};
+		texDesc.Width = _textureSize.x;
+		texDesc.Height = _textureSize.y;
+		texDesc.MipLevels = 1;
+		texDesc.ArraySize = 1;
+		texDesc.Format = DXGI_FORMAT_B8G8R8A8_UNORM;
+		texDesc.SampleDesc.Count = 1;
+		texDesc.Usage = D3D11_USAGE_DYNAMIC;
+		texDesc.BindFlags = D3D11_BIND_SHADER_RESOURCE;
+		texDesc.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+		texDesc.MiscFlags = 0;
+
+		if (FAILED(_d3dDevice->CreateTexture2D(&texDesc, nullptr, &_videoTexture)))
+		{
+			TENLog("Failed to create video texture", LogLevel::Error);
+			return false;
+		}
+
+		if (_videoTexture != nullptr && FAILED(_d3dDevice->CreateShaderResourceView(_videoTexture, nullptr, &_textureView)))
+		{
+			TENLog("Failed to create shader resource view", LogLevel::Error);
+			return false;
+		}
+
+		return true;
+	}
+
+	void VideoHandler::DeInitD3DTexture()
+	{
+		if (_videoTexture)
+		{
+			_videoTexture->Release();
+			_videoTexture = nullptr;
+		}
+
+		if (_textureView)
+		{
+			_textureView->Release();
+			_textureView = nullptr;
+		}
+	}
+
+	void VideoHandler::DeInitPlayer()
+	{
+		if (_player == nullptr)
+			return;
+
+		libvlc_media_player_stop_async(_player);
+		libvlc_media_player_release(_player);
+		_player = nullptr;
+
+		HandleError();
+	}
+
+	void* VideoHandler::OnLockFrame(void* data, void** pixels)
+	{
+		VideoHandler* player = static_cast<VideoHandler*>(data);
+		*pixels = player->_frameBuffer.data();
+		return nullptr;
+	}
+
+	void VideoHandler::OnUnlockFrame(void* data, void* picture, void* const* pixels)
+	{
+		VideoHandler* player = static_cast<VideoHandler*>(data);
+
+		if (!player->_videoTexture)
+		{
+			TENLog("Video texture does not exist", LogLevel::Error);
+			return;
+		}
+
+		// In case of consequential unlock frame callbacks, do not attempt to map texture again.
+		if (player->_needRender)
+			return;
+
+		D3D11_MAPPED_SUBRESOURCE mappedResource;
+		if (SUCCEEDED(player->_d3dContext->Map(player->_videoTexture, 0, D3D11_MAP_WRITE_DISCARD, 0, &mappedResource)))
+		{
+			memcpy(mappedResource.pData, player->_frameBuffer.data(), player->_textureSize.x * player->_textureSize.y * 4);
+			player->_d3dContext->Unmap(player->_videoTexture, 0);
+			player->_needRender = true;
+		}
+		else
+		{
+			TENLog("Failed to map video texture", LogLevel::Error);
+		}
+	}
+
+	void VideoHandler::OnDisplayFrame(void* data, void* picture)
```

Comment:
I guess this is a function that needs to be implemented from Video handler class?

Reply:
Thank you for noticing, I have removed it, as placeholder callbacks are not needed.